design-docs/opmodes.md

```diff
@@ -0,0 +1,543 @@
+}
+```
+
+### OpMode Classes
+
+There are library base classes for different coding styles of routines.  The `OpMode` interface serves as the base interface for all opmodes.  Most users will use either `LinearOpMode` or `PeriodicOpMode` abstract base class implementations of this interface, but this interface enables users to create more customized opmodes while still utilizing the core robot base class opmode-switching implementation.
+
+The full lifecycle of a opmode is as follows:
+- Operator selects opmode on DS -> opmode object is constructed
+- If different opmode is selected -> `opModeClose()` is called (which typically just calls `close()`), object is released to GC
+- While opmode is selected and robot is disabled, `disabledPeriodic()` is called
+- When the robot is enabled in the selected opmode, `opmodeRun()` is called; the result of this is different for different opmode base classes:
+  - `start()` is called once (for `PeriodicOpMode`)
+  - `run()` is called (for `LinearOpMode`), or `periodic()` is called periodically (for `PeriodicOpMode`)
+- When the robot is disabled, `opModeStop()` is called (which results in `end()` being called for `PeriodicOpMode`), followed by `opModeClose()` (which results in `close()` being called for both `PeriodicOpMode` and `LinearOpMode`), object is released to GC
+
+Following `opModeClose()` being called, a *new* opmode object is constructed based on the DS teleop/auto/test/match selector and selected opmode.  In teleop/auto/test, the drop-down selection will be the same as before the previous enable, so the same opmode class is constructed again.  In match (or when FMS-connected), only the selected auto opmode object is initially constructed; once auto completes, the selected teleop opmode object is constructed.  Thus only zero or one opmode objects will ever be "alive" at any given time.
+
+For consistency in operation, the library will ensure that `disabledPeriodic()` is always called at least once before `opModeRun()` is called.
+
+User implementations of opmode classes may have either a no-parameter constructor or a constructor that accepts the user's `Robot` class type.  If available, the library will call the latter and pass the user's `Robot` object to it when constructing the class.
+
+The library will use escalating steps to attempt to terminate a opmode if `opModeRun()` does not return within a reasonable timeframe after `opModeStop()` is called, up to and including termination of the robot executable process (which will result in an automatic restart of it at the system level).  User code (particularly in `LinearOpMode` derived classes) should check for `isRunning()` returning false and return as quickly as possible to allow the opmode to terminate gracefully.
+
+```java
+public interface OpMode {
+  // this function is called periodically while the opmode is selected on the DS (robot is disabled)
+  default void disabledPeriodic() {}
```

Comment:
Since we guarantee this will be called at least once before all other lifecycle methods, this means that this can be called when the robot is enabled (if a new opmode is started while the robot is enabled). ([This comment on the MyTeleop constructor comment](https://github.com/wpilibsuite/allwpilib/pull/7744#discussion_r2360212354) also discusses when a new opmode may start while the robot is enabled.)